cw/util/trace_library.py

```diff
@@ -0,0 +1,110 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+import sqlalchemy as db
+import sqlalchemy_utils as db_utils
+from dataclasses import dataclass, asdict
+from datetime import datetime
+from sqlalchemy.orm import sessionmaker
+from typing import Optional
+
+@dataclass
+class Trace:
+    """ Trace.
+
+    Defines the trace format.
+
+    """
+    wave: bytearray
+    plaintext: bytearray
+    ciphertext: bytearray
+    key: bytearray
+    x_pos: Optional[int] = 0
+    y_pos: Optional[int] = 0
+    trace_id: Optional[int] = 0
+
+class TraceLibrary:
+    """ Class for writing/reading traces to a database.
+
+    Either a new database is created or traces are appended to the existing
+    one. Traces are first written into memory and the flushed into the
+    database after reaching a trace memory threshold.
+    """
+    def __init__(self, db_name, trace_threshold):
+        self.engine = db.create_engine("sqlite:///" + db_name)
+        db_utils.create_database(self.engine.url)
+        self.Session = sessionmaker(self.engine)
+        self.session = self.Session()
+        self.metadata = db.MetaData()
+        self.Traces = db.Table(
+            "traces",
+            self.metadata,
+            db.Column("trace_id", db.Integer, primary_key=True,
+                    autoincrement=True),
+            db.Column("wave", db.LargeBinary),
+            db.Column("plaintext", db.LargeBinary),
+            db.Column("ciphertext", db.LargeBinary),
+            db.Column("key", db.LargeBinary),
+            db.Column("x_pos", db.Integer),
+            db.Column("y_pos", db.Integer),
+        )
+        self.metadata.create_all(self.engine)
+        self.trace_mem = []
+        self.trace_mem_ctr = 0
+        self.trace_mem_thr = trace_threshold
+        self.settingsDict = {}
+
+    def commit_traces(self):
+        """ Writes traces from memory into database.
+        """
+        if len(self.trace_mem):
+            query = db.insert(self.Traces)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
+            self.session.execute(query, traces)
+            self.session.commit()
+            self.trace_mem_ctr = 0
+            self.trace_mem=[]
+
+    def write_trace(self, trace):
+        """ Write traces into database.
+
+        The trace first is written into memory. When reaching the trace
+        memory threshold, the traces are flushed into the database storage.
+
+        Args:
+            trace: The trace to write.
+        """
+        self.trace_mem.append(trace)
+        self.trace_mem_ctr += 1
+        if (self.trace_mem_ctr >= self.trace_mem_thr):
+            self.commit_traces()
+    
+    def get_traces(self, start=0, end=0):
```

Comment:
later, if we start using x and y, we would probably want to have them as optional args here

Reply:
They are already optional. When not providing start and end, all available traces are returned.